flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/JdbcLockFactory.java

```diff
@@ -0,0 +1,246 @@
+  private static final String INIT_LOCK_TABLES_PROPERTY = "maintenance.lock.jdbc.init-lock-tables";
+  private static final String LOCK_TABLE_NAME = "maintenance_lock";
+  private static final int MAX_PREFIX_LENGTH = 100;
+  private static final String CREATE_LOCK_TABLE_SQL =
+      "CREATE TABLE "
+          + LOCK_TABLE_NAME
+          + "(LOCK_TYPE CHAR(1) NOT NULL, ISSUER VARCHAR("
+          + MAX_PREFIX_LENGTH
+          + ") NOT NULL, LOCK_ID CHAR(36) NOT NULL, PRIMARY KEY (LOCK_TYPE, ISSUER))";
+  private static final String CREATE_LOCK_SQL =
+      "INSERT INTO " + LOCK_TABLE_NAME + " (LOCK_TYPE, ISSUER, LOCK_ID) VALUES (?, ?, ?)";
+  private static final String GET_LOCK_SQL =
+      "SELECT LOCK_ID FROM " + LOCK_TABLE_NAME + " WHERE LOCK_TYPE=? AND ISSUER=?";
+  private static final String DELETE_LOCK_SQL =
+      "DELETE FROM " + LOCK_TABLE_NAME + " WHERE LOCK_TYPE=? AND ISSUER=? AND LOCK_ID=?";
+
+  private final String uri;
+  private final String issuer;
+  private final Map<String, String> properties;
+  private transient JdbcClientPool pool;
+
+  public JdbcLockFactory(String uri, String issuer, Map<String, String> properties) {
+    Preconditions.checkNotNull(uri, "JDBC connection URI is required");
+    Preconditions.checkArgument(
+        issuer.length() < MAX_PREFIX_LENGTH,
+        "Invalid prefix length: issuer should be shorter than %s",
+        MAX_PREFIX_LENGTH);
+    this.uri = uri;
+    this.issuer = issuer;
+    this.properties = properties;
+  }
+
+  @Override
+  public void open() {
+    this.pool = new JdbcClientPool(1, uri, properties);
+
+    if (PropertyUtil.propertyAsBoolean(properties, INIT_LOCK_TABLES_PROPERTY, false)) {
+      initializeLockTables();
+    }
+  }
+
+  /** Only used in testing to share the jdbc pool */
+  @VisibleForTesting
+  void open(JdbcLockFactory other) {
+    this.pool = other.pool;
+  }
+
+  @Override
+  public Lock createLock() {
+    return new Lock(pool, issuer, Type.MAINTENANCE);
+  }
+
+  @Override
+  public Lock createRecoveryLock() {
+    return new Lock(pool, issuer, Type.RECOVERY);
+  }
+
+  @Override
+  public void close() throws IOException {
+    pool.close();
+  }
+
+  private void initializeLockTables() {
+    LOG.trace("Creating database tables (if missing) to store table maintenance locks");
+    try {
+      pool.run(
+          conn -> {
+            DatabaseMetaData dbMeta = conn.getMetaData();
+            ResultSet tableExists =
+                dbMeta.getTables(
+                    null /* catalog name */,
+                    null /* schemaPattern */,
+                    LOCK_TABLE_NAME /* tableNamePattern */,
+                    null /* types */);
+            if (tableExists.next()) {
+              return true;
+            }
+
+            LOG.debug("Creating table {} to store iceberg catalog tables", LOCK_TABLE_NAME);
+            return conn.prepareStatement(CREATE_LOCK_TABLE_SQL).execute();
```

Comment:
check return value and log the execution result (success/failure)

Reply:
I don't see when this could be a failure and there would be no exception thrown. We could change the return null instead of the output of the execute, as we need to return something from the `run` method anyway, but I don't see a point adding more code here than neccessary.

WDYT?

Reply:
my mistake. I thought the boolean return value from `execute` is success or failure. it is not.
```
true if the first result is a ResultSet object; false if the first result is an update count or there is no result
```